pkg/gitops/rollback.go

```diff
@@ -0,0 +1,210 @@
+	message := "Rollback to commit " + targetCommit.Hash.String()
+	if err := cr.AddAndCommit(username, email, message); err != nil {
+		return fmt.Errorf("error while committing rollback: %w", err)
+	}
+	cr.RollbackMode = false
+	klog.V(2).InfoS("Rollback successful", "targetCommit", targetCommit.Hash.String())
+	return nil
+}
+
+func resetWorktree(w *git.Worktree, hash plumbing.Hash, mode git.ResetMode) error {
+	options := &git.ResetOptions{
+		Commit: hash,
+		Mode:   mode,
+	}
+	if err := w.Reset(options); err != nil {
+		return fmt.Errorf("unable to reset worktree")
+	}
+	return nil
+}
+
+func (cr *CustomRepo) doDeletePatch(patch *object.Patch) error {
+	for _, filePatch := range patch.FilePatches() {
+		fromFile, toFile := filePatch.Files()
+		if toFile == nil {
+			path := filepath.Join(cr.Dir, fromFile.Path())
+			resource, err := cr.getResourceByPath(path)
+			if err != nil {
+				return fmt.Errorf("unable to read resource at path %s: %w", path, err)
+			}
+			if err := cr.K8s.DeleteResource(resource); err != nil {
+				return fmt.Errorf("unable to delete resource %s: %w", resource.GetName(), err)
+			}
+			klog.V(2).InfoS("(Rollback) Deleted file", "path", path)
+		}
+	}
+	return nil
+}
+
+func (cr *CustomRepo) doCreateUpdatePatch(patch *object.Patch) error {
+	for _, filePatch := range patch.FilePatches() {
+		_, toFile := filePatch.Files()
+		if toFile != nil {
+			path := filepath.Join(cr.Dir, toFile.Path())
+			resource, err := cr.getResourceByPath(path)
+			if err != nil {
+				return fmt.Errorf("unable to read resource at path %s: %w", path, err)
+			}
+			if err := cr.K8s.CreateOrUpdateResource(resource); err != nil {
+				return fmt.Errorf("unable to create/update resource %s: %w", resource.GetName(), err)
+			}
+			klog.V(2).InfoS("(Rollback) Created/Updated file", "path", path)
+		}
+	}
+	return nil
+}
+
+func (cr *CustomRepo) getResourceByPath(path string) (*unstructured.Unstructured, error) {
+	resource := &unstructured.Unstructured{}
+	gvk := schema.GroupVersionKind{}
+	if err := cr.readResource(resource, path); err != nil {
+		return nil, fmt.Errorf("unable to read resource: %w", err)
+	}
+	apiVersion := resource.GetAPIVersion()
+	kind := resource.GetKind()
+	if apiVersion == "networking.k8s.io/v1" {
+		gvk.Group = "networking.k8s.io"
+		gvk.Version = "v1"
+	} else if apiVersion == "crd.antrea.io/v1alpha1" {
+		gvk.Group = "crd.antrea.io"
+		gvk.Version = "v1alpha1"
+	} else {
+		return nil, fmt.Errorf("unknown apiVersion found: %s", apiVersion)
+	}
+	gvk.Kind = kind
+	resource.SetGroupVersionKind(gvk)
+	return resource, nil
+}
+
+func (cr *CustomRepo) readResource(resource *unstructured.Unstructured, path string) error {
+	var y []byte
+	if cr.StorageMode == StorageModeDisk {
+		y, _ = ioutil.ReadFile(path)
+	} else {
+		fstat, _ := cr.Fs.Stat(path)
+		y = make([]byte, fstat.Size())
+		f, err := cr.Fs.Open(path)
+		if err != nil {
+			return fmt.Errorf("error opening file")
+		}
+		f.Read(y)
+	}
```

Comment:
question: I thought `cr.Fs` could be used to abstract away the fact that the repo can be on disk or in-memory? In other words, shouldn't something like `cr.Fs.Open` work in both cases and could we have unified code that way (no `if cr.StorageMode == StorageModeDisk`)?

Reply:
For the in-memory case, we perform operations on the filesystem object stored in cr.Fs, whereas for the on-disk case, we call functions from the os and ioutil libraries to read/modify the disk file system directly. Unless there's a way to store the disk file system object as a filesystem object like we do for the in-memory case, it wasn't obvious to us how to abstract operations like you say - any ideas?

Reply:
On second thought, maybe we can create our own custom filesystem object with Open, Mkdir functions that handle the StorageModeDisk vs. StorageModeInMemory logic so we don't have to worry about the underlying filesystem when we call these functions, like you said

Reply:
No, that's not what I meant. I meant something like this (for initialization), although I have not tested it:

```go
	var storer git.Storer
	var fs billy.Filesystem
	if StorageMode == StorageModeDisk {
		fs = osfs.New(<DIR>)
		storer = filesystem.NewStorage(fs, cache.NewObjectLRUDefault())
 
	} else {
		storer = memory.NewStorage()
		fs = memfs.New()
	}
	r, err := git.Init(storer, fs) // same Init call for both modes
```
After that you use the `fs` object (of type `billy.Filesystem`) for any operation you need to do: create a directory, remove a file, etc. It doesn't matter whether it's in memory or on disk.
I'll merge this PR now, but please open a follow-up PR to make a change like this one if possible.

Reply:
I see what you mean - will try this now